src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7356,9 +7356,9 @@ private StatementSyntax ParseEmbeddedStatement(bool complexCheck)
                 case SyntaxKind.LocalFunctionStatement:
                     statement = this.AddError(statement, ErrorCode.ERR_BadEmbeddedStmt);
                     break;
-                // Expression statements as embedded statements should be followed by semicolon.
+                // Expression statements as embedded statements should be followed by semicolon in scripts.
                 case SyntaxKind.ExpressionStatement:
-                    if (((ExpressionStatementSyntax)statement).SemicolonToken.IsMissing)
+                    if (IsScript && ((ExpressionStatementSyntax)statement).SemicolonToken.IsMissing)
```

Comment:
This condition is not sufficient to ensure that an error has not already been reported. In script scenarios where the compiler found a semicolon missing in the middle of the code, this may cause a duplicate error to be produced.

May I suggest that you check that the semicolon token doesn't already have an error on it?